compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp

```diff
@@ -0,0 +1,1364 @@
+
+    for (auto it : llvm::enumerate(llvm::ArrayRef<bool>(toInsertDims))) {
+      if (!it.value()) reassociationMap.push_back({});
+      reassociationMap.back().push_back(
+          rewriter.getAffineDimExpr(it.index() + 1));
+    }
+
+    llvm::SmallVector<Value> expandedUpdates;
+    for (auto update : op.getUpdates()) {
+      auto updatesTy = update.getType().cast<ShapedType>();
+
+      llvm::SmallVector<int64_t> newShape;
+      for (int i = 0, s = reassociationMap.size(); i < s; ++i) {
+        newShape.push_back(updatesTy.getDimSize(i));
+        for (int j = 1, s = reassociationMap[i].size(); j < s; ++j) {
+          newShape.push_back(1);
+        }
+      }
+
+      Value expandUpdate = rewriter.create<tensor::ExpandShapeOp>(
+          op.getLoc(),
+          RankedTensorType::get(newShape, updatesTy.getElementType()), update,
+          reassociationMap);
+      expandedUpdates.push_back(expandUpdate);
+    }
+
+    llvm::SmallVector<int64_t> newUpdatedWindowDims(toInsertDims.size());
+    llvm::SmallVector<int64_t> newInsertedWindowDims(frontInsertedDims);
+    std::iota(newUpdatedWindowDims.begin(), newUpdatedWindowDims.end(), 1);
+    std::iota(newInsertedWindowDims.begin(), newInsertedWindowDims.end(), 0);
+
+    auto newDimNumbers = mlir::stablehlo::ScatterDimensionNumbersAttr::get(
+        op.getContext(), newUpdatedWindowDims, newInsertedWindowDims,
+        dimNumbers.getScatterDimsToOperandDims(),
+        /*indexVectorDim=*/1);
+
+    auto newScatter = rewriter.create<mlir::stablehlo::ScatterOp>(
+        op.getLoc(), op.getResultTypes(), op.getInputs(),
+        op.getScatterIndices(), expandedUpdates, newDimNumbers,
+        op.getIndicesAreSorted(), op.getUniqueIndices());
+    Region &region = newScatter.getUpdateComputation();
+    rewriter.cloneRegionBefore(op.getUpdateComputation(), region, region.end());
+    rewriter.replaceOp(op, newScatter.getResults());
+    return success();
+  }
+};
+
+// Traverse upward past common operations to see if the value came from a
+// boolean tensor.
+bool isFromBool(Value val) {
+  while (true) {
+    Operation *op = val.getDefiningOp();
+    if (!op) return false;
+
+    if (auto convertOp = dyn_cast<mlir::stablehlo::ConvertOp>(op)) {
+      auto inTy = convertOp.getOperand().getType().cast<ShapedType>();
+      if (inTy.getElementType().isInteger(1)) {
+        return true;
+      }
+      val = convertOp.getOperand();
+      continue;
+    }
+
+    if (isa<mlir::stablehlo::DynamicBroadcastInDimOp>(op) ||
+        isa<mlir::stablehlo::BroadcastInDimOp>(op) ||
+        isa<mlir::stablehlo::BroadcastOp>(op)) {
+      val = op->getOperand(0);
+      continue;
+    }
+
+    return false;
+  }
+}
+
+// StableHLO of non-finite values (e.g. NaN, inf) and 0.0 produce 0.0 for XLA.
+// For linalg we need to convert these to select operations.
+struct MulCastOfBool final : OpRewritePattern<mlir::stablehlo::MulOp> {
+  using OpRewritePattern::OpRewritePattern;
```

Comment:
"for XLA" comment still useful in IREE?

Reply:
Thanks, good catch. Actually I couldn't parse the whole comment at all, I think they wanted to say something along the lines: 'Mul of .... produces 0.0 in XLA'.